tensorflow/compiler/mlir/hlo/lib/Dialect/mhlo/transforms/hlo_canonicalize_reduction.cc

```diff
@@ -0,0 +1,238 @@
+//       -> tensor<?x?f32> return %3 : tensor<?x?xf32>
+//     }
+//  ```
+//
+// For case b):
+// ====================================================================================
+//   we convert all row reduction to rank-2 row reduction.
+//   For example, suppose we have:
+//   ```
+//     func @test(%arg0: tensor<?x?x?xf32>) -> tensor<?x?xf32> {
+//       ...
+//       %2 = "mhlo.reduce"(%arg0, ...) ( {...})
+//         {dimensions = dense<[2]> : tensor<1xi64>} :
+//         (tensor<?x?x?xf32>, tensor<f32>) -> tensor<?x?xf32>
+//       return %2 : tensor<?x?xf32>
+//     }
+//  ```
+//   After conversion:
+//     func @test(%arg0: tensor<?x?x?xf32>) -> tensor<?x?xf32> {
+//       // [a, b, c] -> [a*b, c]
+//       %1 = mhlo.dynamic_reshape(%arg0, ...) : (tensor<?x?x?xf32>,
+//       tensor<2xi64>) -> tensor<?x?xf32> %2 = "mhlo.reduce"(%1, ...) ( {...})
+//         {dimensions = dense<[1]> : tensor<1xi64>} :
+//         (tensor<?x?xf32>, tensor<f32>) -> tensor<?xf32>
+//       %3 = "mhlo.dynamic_reshape"(%2, ...) : (tensor<?xf32>, tensor<1xi64>)
+//       -> tensor<?x?f32> return %3 : tensor<?x?xf32>
+//     }
+//  ```
+//
+// For case c):
+// ====================================================================================
+//   we convert all reduce-to-scalar to rank-2 column reduction.
+//
+//   For example, suppose we have:
+//   ```
+//     func @test(%arg0: tensor<?x?x?xf32>) -> tensor<f32> {
+//       ...
+//       %2 = "mhlo.reduce"(%arg0, ...) ( {...})
+//         {dimensions = dense<[0,1,2]> : tensor<3xi64>} :
+//         (tensor<?x?x?xf32>, tensor<f32>) -> tensor<f32>
+//       return %2 : tensor<f32>
+//     }
+//  ```
+//   After conversion:
+//     func @test(%arg0: tensor<?x?x?xf32>) -> tensor<f32> {
+//       // [a, b, c] -> [a*b*c, 1]
+//       %1 = mhlo.dynamic_reshape(%arg0, ...) : (tensor<?x?x?xf32>,
+//       tensor<2xi64>) -> tensor<?x?xf32> %2 = "mhlo.reduce"(%1, ...) ( {...})
+//         {dimensions = dense<[0]> : tensor<1xi64>} :
+//         (tensor<?x?xf32>, tensor<f32>) -> tensor<?xf32>
+//       %3 = "mhlo.reshape"(%2, ...) : (tensor<?xf32>, tensor<1xi64>) ->
+//       tensor<f32> return %3 : tensor<f32>
+//     }
+//  ```
+
+struct HloCanonicalizeReductionPass
+    : HloCanonicalizeReductionPassBase<HloCanonicalizeReductionPass> {
+  void runOnFunction() override {
+    getFunction().walk([&](ReduceOp op) {
+      SmallVector<int64_t, 4> dims_to_reduce;
+      DenseSet<int64_t> dims_to_reduce_set;
+      for (auto dim : op.dimensions().getIntValues()) {
+        dims_to_reduce.push_back(dim.getSExtValue());
+        dims_to_reduce_set.insert(dims_to_reduce.back());
+      }
+
+      // empty reduction is just a no-op, thus no need to do codegen.
+      if (dims_to_reduce.empty()) return;
+      auto ty = op.getOperand(0).getType().dyn_cast<RankedTensorType>();
+      assert(ty && ty.hasRank() && "suppose reduce input is a ranked tensor");
```

Comment:
This should either be part of the verifier for `ReduceOp` (in which case you can remove the assert) or it isn't (in which case an assert isn't suitable and we should fail gracefully).

Also, `ty.hasRank()` is always true on a `RankedTensorType`.

Reply:
ok, I check the `ty` and `return signalPassFailure();` instead.